README.md
# NAME

MooseX::MojoControllerExposingAttributes - Expose controller attributes to Mojolicious

# VERSION

version 1.000001

# SYNOPSIS

    package MyApp::Controller::Example;
    use MooseX::MojoControllerExposingAttributes;

    ...;

    has some_attribute => (
        is     => 'ro',
        traits => ['ExposeMojo'],
    );

    # then later in a template: <%= ctrl->some_attribute %>

# DESCRIPTION

This class allows you to expose _selected_ Moose attributes from your
Mojolicious controller to your templates by marking them with the `ExposeMojo`
trait.

Using this class in a Perl class does several things:

- It makes the class a subclass of Mojolicious::Controller
- It sets up the class with Moose and Moose::NonMoose
- It applies the extra role and metaclass traits to the class so this works with [Mojolicious::Plugin::ExposeControllerMethod](https://metacpan.org/pod/Mojolicious::Plugin::ExposeControllerMethod)
- It sets up the `ExposeMojo` trait

So rather than declaring your controller class a Moose Mojolicious Controller in
the usual way:

    package MyApp::Controller::Example;
    use Mojo::Base 'Mojolicious::Controller';
    use Moose::NonMoose;
    use Moose;

You should simply say:

    package MyApp::Controller::Example;
    use MooseX::MojoControllerExposingAttributes;

Once you've done that then you can define attributes in the class (or in roles
the class consumes) that are exposed to Mojolicious.

    has some_attribute => (
        is     => 'ro',
        traits => ['ExposeMojo'],
    );

    has some_attribute_with_a_really_long_name => (
       is                => 'ro',
       traits            => ['ExposeMojo'],
       expose_to_mojo_as => 'shorter_name',
    );

In order to get the `ctrl` helper you should make sure you've loaded the
[Mojolicious::Plugin::ExposeControllerMethod](https://metacpan.org/pod/Mojolicious::Plugin::ExposeControllerMethod) plugin somewhere in your
Mojolicious application, typically within the `startup` method itself:

    sub startup {
        my $self = shift;

        $self->plugin('ExposeControllerMethod');

        ...
    }

Then you'll be able to access your attributes from within templates that
are rendered from that controller:

    some attribute: <%= ctrl->some_attribute %>
    some attribute with a really long name: <%= ctrl->shorter_name %>

# BUGS

It would be nice to be able to set the baseclass instead of always
using Mojolicious::Controller

# SEE ALSO

[Mojolicious::Plugin::ExposeControllerMethod](https://metacpan.org/pod/Mojolicious::Plugin::ExposeControllerMethod)

[MooseX::MojoControllerExposingAttributes::Trait::Attribute](https://metacpan.org/pod/MooseX::MojoControllerExposingAttributes::Trait::Attribute)

# AUTHOR

Mark Fowler <mfowler@maxmind.com>

# CONTRIBUTORS

- Dave Rolsky <drolsky@maxmind.com>
- Olaf Alders <oalders@maxmind.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2015 - 2016 by MaxMind, Inc.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
{{$NEXT}}

1.000001 2016-03-21
    - Pod updates

1.000000 2016-03-15
    - Initial release


t/my_app/lib/MyApp/Controller/Example.pm
package MyApp::Controller::Example;
use MooseX::MojoControllerExposingAttributes;

has name => (
    is => 'ro',
    lazy => 1,
    default => 'Mark Fowler',
    traits => ['ExposeMojo'],
);

has any_other_name => (
    is => 'ro',
    lazy => 1,
    default => 'Still smells sweet',
    traits => ['ExposeMojo'],
    expose_to_mojo_as => 'rose',
);

has custom => (
    is => 'ro',
    lazy => 1,
    default => 'custom reader works',
    traits => ['ExposeMojo'],
    reader => 'get_custom',
);

# This action will render a template
sub welcome {
    my $self = shift;

    # Render template "example/welcome.html.ep" with message
    $self->render(
        msg => 'Welcome to the Mojolicious real-time web framework!',
    );
}

1;


lib/MooseX/MojoControllerExposingAttributes.pm
## no critic (Moose::RequireMakeImmutable)
package MooseX::MojoControllerExposingAttributes;

use strict;
use warnings;

# ensure loaded, even but don't have the side effects on this class
use Moose            ();
use MooseX::NonMoose ();

# modules we use ourselves
use Moose::Exporter;
use MooseX::MojoControllerExposingAttributes::Trait::Attribute;
use MooseX::MojoControllerExposingAttributes::Trait::Class;
use Import::Into;

our $VERSION = '1.000002';

my $baseclass = 'Mojolicious::Controller';

# While the unimport / init_meta installation this creates is fine to install
# directly, we need a customer importer

my ($import) = Moose::Exporter->build_import_methods(

    # give the class the role that has the 'controller_method_name' method in it
    base_class_roles =>
        ['MooseX::MojoControllerExposingAttributes::Role::Class'],

    # this allows the meta class to have 'get_attribute_for_mojo_helper' which
    # is used by 'controller_method_name' to do the lookup
    class_metaroles => {
        class => ['MooseX::MojoControllerExposingAttributes::Trait::Class'],
    },

    install => [qw(unimport init_meta)],
);

sub import {
    my $target = caller;

    # do the Moose/NonMoose dance
    # TODO, shouldn't we do something with unimport also so we can "no Moose" ?
    MooseX::NonMoose->import::into($target);
    Moose->import::into($target);

    # set the baseclass
    # this must happen AFTER we've moosified the package, but BEFORE the
    # M::E import routine monkeys around with the metaclass (a side effect
    # of how MooseX::NonMoose does its magic)
    $target->meta->superclasses($baseclass);

    # now do the normal import
    goto &$import;
}

1;

# ABSTRACT: Expose controller attributes to Mojolicious

__END__

=pod

=head1 SYNOPSIS

    package MyApp::Controller::Example;
    use MooseX::MojoControllerExposingAttributes;

    ...;

    has some_attribute => (
        is     => 'ro',
        traits => ['ExposeMojo'],
    );

    # then later in a template: <%= ctrl->some_attribute %>

=head1 DESCRIPTION

This module is for advanced use.  C<$c>/C<$self> are already made available in
templates and are likely sufficient for the majority of use cases.  This module
was created in order to expose L<Moose> attributes in a way where you don't
have to stash them every single time you want to use them.

This class allows you to expose I<selected> Moose attributes from your
Mojolicious controller to your templates by marking them with the C<ExposeMojo>
trait.

Using this class in a Perl class does several things:

=over

=item It makes the class a subclass of Mojolicious::Controller

=item It sets up the class with Moose and Moose::NonMoose

=item It applies the extra role and metaclass traits to the class so this works with L<Mojolicious::Plugin::ExposeControllerMethod>

=item It sets up the C<ExposeMojo> trait

=back

So rather than declaring your controller class a Moose Mojolicious Controller in
the usual way:

   package MyApp::Controller::Example;
   use Mojo::Base 'Mojolicious::Controller';
   use Moose::NonMoose;
   use Moose;

You should simply say:

   package MyApp::Controller::Example;
   use MooseX::MojoControllerExposingAttributes;

Once you've done that then you can define attributes in the class (or in roles
the class consumes) that are exposed to Mojolicious.

  has some_attribute => (
      is     => 'ro',
      traits => ['ExposeMojo'],
  );

  has some_attribute_with_a_really_long_name => (
     is                => 'ro',
     traits            => ['ExposeMojo'],
     expose_to_mojo_as => 'shorter_name',
  );

In order to get the C<ctrl> helper you should make sure you've loaded the
L<Mojolicious::Plugin::ExposeControllerMethod> plugin somewhere in your
Mojolicious application, typically within the C<startup> method itself:

    sub startup {
        my $self = shift;

        $self->plugin('ExposeControllerMethod');

        ...
    }

Then you'll be able to access your attributes from within templates that
are rendered from that controller:

   some attribute: <%= ctrl->some_attribute %>
   some attribute with a really long name: <%= ctrl->shorter_name %>

=head1 BUGS

It would be nice to be able to set the baseclass instead of always
using Mojolicious::Controller

=head1 SEE ALSO

L<Mojolicious::Plugin::ExposeControllerMethod>

L<MooseX::MojoControllerExposingAttributes::Trait::Attribute>

=cut


lib/MooseX/MojoControllerExposingAttributes/Trait/Class.pm
package MooseX::MojoControllerExposingAttributes::Trait::Class;
use Moose::Role;

use MooseX::Types::Moose qw( HashRef );

our $VERSION = '1.000002';

has _mojo_method_name_to_attribute_reader_name_map => (
    is      => 'ro',
    isa     => HashRef,
    lazy    => 1,
    handles => { _get_mojo_attribute_name_for_method => 'get' },
    builder => '_make_mojo_method_name_to_attribute_reader_name_map',
);

# this isn't called _build_XXX because we're also calling it directly too
sub _make_mojo_method_name_to_attribute_reader_name_map {
    my $self = shift;

    my %output;
    foreach
        my $attr ( sort { $a->name cmp $b->name } $self->get_all_attributes )
    {
        next
            unless $attr->can('does')
            && $attr->does(
            'MooseX::MojoControllerExposingAttributes::Trait::Attribute');
        $output{ $attr->expose_to_mojo_as || $attr->name }
            = $attr->get_read_method;
    }
    return \%output;
}

sub get_read_method_name_for_mojo_helper {
    my $self        = shift;
    my $wanted_name = shift;

    return $self->_get_mojo_attribute_name_for_method($wanted_name)
        if $self->is_immutable;
    return
        $self->_make_mojo_method_name_to_attribute_reader_name_map
        ->{$wanted_name};
}

no Moose::Role;
1;

=head1 NAME

MooseX::MojoControllerExposingAttributes::Trait::Class - metaclass helper for MooseX::MojoControllerExposingAttributes

=head1 SYNOPSIS

    # No user serviceable parts contained within

=head1 DESCRIPTION

This is a class trait that is applied to your class's metaclass by
L<MooseX::MojoControllerExposingAttributes>.

You probably don't want to be worried about the internal details of this.

=head1 METHODS

=head2 get_read_method_name_for_mojo_helper( $wanted_name )

Returns the name of the reader method for the attribute exposed to Mojolicious
as the passed argument, or, if no such attribute is exposed, the undefined
value.

This method works by examining each attribute to see if the attribute
has the correct trait and what name the attribute is exposed to Mojolicious as.
If the metaclass has been made immutable this examination only has to happen
once and the results are cached.

=head1 SEE ALSO

L<MooseX::MojoControllerExposingAttributes>


lib/MooseX/MojoControllerExposingAttributes/Trait/Attribute.pm
package MooseX::MojoControllerExposingAttributes::Trait::Attribute;
use Moose::Role;
Moose::Util::meta_attribute_alias('ExposeMojo');

our $VERSION = '1.000002';

use MooseX::Types::Moose qw( Str );

has expose_to_mojo_as => (
    is  => 'ro',
    isa => Str,    # todo, this should be a method name constraint
);

no Moose::Role;
1;
__END__

=head1 NAME

MooseX::MojoControllerExposingAttributes::Trait::Attribute - trait used to expose attribute to Mojolicious

=head1 SYNOPSIS

    package MyApp::Controller::Example;
    use MooseX::MojoControllerExposingAttributes;

    ...;

    has some_attribute => (
        is     => 'ro',
        traits => ['ExposeMojo'],
    );

    # then later in a template: <%= ctrl->some_attribute %>

=head1 DESCRIPTION

This class is an attribute trait that can be applied with C<ExposeMojo>.

Applying this trait to an attribute within a L<Mojolicious::Controller> subclass
allows reading of that attribute from within a Mojolicious template by calling
the corresponding method name on the C<ctrl> helper.

In order to this to work you must also correctly setup the metaclass of the
controller (usually by simply doing a
C<use MooseX::MojoControllerExposingAttributes;> within that controller class)
and load the L<Mojolicious::Plugin::ExposeControllerMethod> plugin within your
Mojolicious application to provide the C<ctrl> helper.

=head1 ATTRIBUTES

=head2 expose_to_mojo_as

Provide an alternative name this attribute should be exposed as

    # expose the rose attribute so it can be called via ctrl->other_name
    # rather than ctrl->rose
    has rose => (
        is                => 'ro',
        traits            => ['ExposeMojo'],
        expose_to_mojo_as => 'other_name',
    );

By default if C<expose_to_mojo_as> isn't provided the attribute will be exposed
as a method with the same name as the attribute (worth noting: this is not
necessarily the same as the name of the reader method that you'd normally use to
access the attribute if you've defined that to be different to the attribute
name.)

=head1 SEE ALSO

L<Mojolicious::Plugin::ExposeControllerMethod>

L<MooseX::MojoControllerExposingAttributes>


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.043.
use strict;
use warnings;

use 5.013010;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Expose controller attributes to Mojolicious",
  "AUTHOR" => "Mark Fowler <mfowler\@maxmind.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "MooseX-MojoControllerExposingAttributes",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.013010",
  "NAME" => "MooseX::MojoControllerExposingAttributes",
  "PREREQ_PM" => {
    "Import::Into" => 0,
    "Mojolicious::Plugin::ExposeControllerMethod" => "1.000000",
    "Moose" => 0,
    "Moose::Exporter" => 0,
    "Moose::Role" => 0,
    "MooseX::NonMoose" => 0,
    "MooseX::Types::Moose" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Basename" => 0,
    "File::Spec" => 0,
    "File::Spec::Functions" => 0,
    "Mojo::Base" => 0,
    "Test::Mojo" => 0,
    "Test::More" => "0.96",
    "lib" => 0
  },
  "VERSION" => "1.000002",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Basename" => 0,
  "File::Spec" => 0,
  "File::Spec::Functions" => 0,
  "Import::Into" => 0,
  "Mojo::Base" => 0,
  "Mojolicious::Plugin::ExposeControllerMethod" => "1.000000",
  "Moose" => 0,
  "Moose::Exporter" => 0,
  "Moose::Role" => 0,
  "MooseX::NonMoose" => 0,
  "MooseX::Types::Moose" => 0,
  "Test::Mojo" => 0,
  "Test::More" => "0.96",
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
